src/nvim/context.c

```diff
@@ -0,0 +1,383 @@
+
+  if (flags & kCtxRegs) {
+    ctx_restore_regs(ctx);
+  }
+
+  if (flags & kCtxJumps) {
+    ctx_restore_jumps(ctx);
+  }
+
+  if (flags & kCtxBuflist) {
+    ctx_restore_buflist(ctx);
+  }
+
+  if (flags & kCtxGVars) {
+    ctx_restore_gvars(ctx);
+  }
+
+  if (flags & kCtxFuncs) {
+    ctx_restore_funcs(ctx);
+  }
+
+  if (free_ctx) {
+    ctx_free(ctx);
+  }
+
+  set_option_value("shada", 0L, (char *)op_shada, OPT_GLOBAL);
+  xfree(op_shada);
+
+  return true;
+}
+
+/// Saves the global registers to a context.
+///
+/// @param  ctx    Save to this context.
+static inline void ctx_save_regs(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  msgpack_sbuffer_init(&ctx->regs);
+  shada_encode_regs(&ctx->regs);
+}
+
+/// Restores the global registers from a context.
+///
+/// @param  ctx   Restore from this context.
+static inline void ctx_restore_regs(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  shada_read_sbuf(&ctx->regs, kShaDaWantInfo | kShaDaForceit);
+}
+
+/// Saves the jumplist to a context.
+///
+/// @param  ctx  Save to this context.
+static inline void ctx_save_jumps(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  msgpack_sbuffer_init(&ctx->jumps);
+  shada_encode_jumps(&ctx->jumps);
+}
+
+/// Restores the jumplist from a context.
+///
+/// @param  ctx  Restore from this context.
+static inline void ctx_restore_jumps(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  shada_read_sbuf(&ctx->jumps, kShaDaWantInfo | kShaDaForceit);
+}
+
+/// Saves the buffer list to a context.
+///
+/// @param  ctx  Save to this context.
+static inline void ctx_save_buflist(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  msgpack_sbuffer_init(&ctx->buflist);
+  shada_encode_buflist(&ctx->buflist);
+}
+
+/// Restores the buffer list from a context.
+///
+/// @param  ctx  Restore from this context.
+static inline void ctx_restore_buflist(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  shada_read_sbuf(&ctx->buflist, kShaDaWantInfo | kShaDaForceit);
+}
+
+/// Saves global variables to a context.
+///
+/// @param  ctx  Save to this context.
+static inline void ctx_save_gvars(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  msgpack_sbuffer_init(&ctx->gvars);
+  shada_encode_gvars(&ctx->gvars);
+}
+
+/// Restores global variables from a context.
+///
+/// @param  ctx  Restore from this context.
+static inline void ctx_restore_gvars(Context *ctx)
+  FUNC_ATTR_NONNULL_ALL
+{
+  shada_read_sbuf(&ctx->gvars, kShaDaWantInfo | kShaDaForceit);
+}
+
+/// Saves functions to a context.
+///
+/// @param  ctx         Save to this context.
+/// @param  scriptonly  Save script functions only.
+static inline void ctx_save_funcs(Context *ctx, bool scriptonly)
+  FUNC_ATTR_NONNULL_ALL
+{
+  ctx->funcs = (Array)ARRAY_DICT_INIT;
+  Error err = ERROR_INIT;
+
+  HASHTAB_ITER(&func_hashtab, hi, {
+    const char_u *const name = hi->hi_key;
+    bool islambda = (STRNCMP(name, "<lambda>", 8) == 0);
+    bool isscript = (name[0] == K_SPECIAL);
+
+    if (!islambda && (!scriptonly || isscript)) {
```

Comment:
`(!scriptonly || isscript)` seems like a weird condition.  With `scriptonly==true` I would expect it to check `isscript==true`.

Reply:
Hmmmm, well, this part is just equivalent to `(!scriptonly || (scriptonly && isscript))`